src/libYARP_os/src/yarp/os/NetworkClock.cpp

```diff
@@ -111,7 +112,11 @@ bool NetworkClock::Private::read(ConnectionReader& reader)
 
     listMutex.lock();
     auto waiter_it = waiters->begin();
-    while (waiter_it != waiters->end()) {
+    // Update the wake-up time. In case of a time reset it closes the gap between the waiter and _time.
+    waiter_it->first = _time + (waiter_it->first - oldTime);
```

Comment:
If this isn't done only in case of time reset, you will never wake up.
For example
* At T=1 the thread asks for a delay of 2 seconds => `waiter_it->first = 3`
* At T=2 you change `waiter_it->first` to be `_time` (2) + `waiter_it->first - oldTime` (3 - 1 = 2) => `waiter_it->first = 4`
* At T=3 you change `waiter_it->first` to be `_time` (3) + `waiter_it->first - oldTime` (4 - 2 = 2) => `waiter_it->first = 5`

And so on...

This update should be done only if `oldTime > _time`

Reply:
@drdanz thanks for the explanation! 

> This update should be done only if `oldTime > _time`

Is this trigger ok if the clock is non-monotonic?

Reply:
I'd say this is the only check that we can make, we have to assume that the clock might not be monotonic, therefore, if we go back in time, the old time will be bigger than the current one. We don't even care if `oldTime == _time`, because that will leave the wake up time unchanged...

Reply:
I agree with @drdanz, the all idea of "network clock reset" just means "clock that goes backward in time" and so is basically a synonymous of "non-monotonic clocks". Also for non simulated clocks, I think that there are two main cases that should be covered by the fix as proposed by @drdanz : 
* Time reset offset that is much longer then the delay period: think a jump backward in time of 1 hour due to switch from/to daylight savings time, and a thread that has a delay of 1 second: in that case we don't want the thread to wait for 1 hour, and this use case was covered both by the logic of unblocking all the threads, and by the new logic of @drdanz 
* Time reset is shorter then the delay period: think of a small jump backward of a few milliseconds due to time synchronization, if we have a thread that has a delay of 60 seconds (the default for example for RFModule) we don't want the thread to be waked up, and this is the reason of the new logic proposed by @drdanz that you imlemented

In both cases, it is necessary to handle the jump backward in time even if the clock is non-simulated.